# VencoPy Tutorial 2

This tutorial aims to give a more in depth overview into the DataParser class and showcases some features that can be customised.

In [17]:
from pathlib import Path

from vencopy.core.dataparsers import parse_data
from vencopy.utils.utils import load_configs, create_output_folders

In [18]:
base_path = Path.cwd().parent
configs = load_configs(base_path)
create_output_folders(configs=configs)
    
# Adapt relative paths in config for tutorials
configs['dev_config']['global']['relative_path']['parse_output'] = Path.cwd().parent / configs['dev_config']['global']['relative_path']['parse_output']
configs['dev_config']['global']['relative_path']['diary_output'] = Path.cwd().parent / configs['dev_config']['global']['relative_path']['diary_output']
configs['dev_config']['global']['relative_path']['grid_output'] = Path.cwd().parent / configs['dev_config']['global']['relative_path']['grid_output']
configs['dev_config']['global']['relative_path']['flex_output'] = Path.cwd().parent / configs['dev_config']['global']['relative_path']['flex_output']
configs['dev_config']['global']['relative_path']['aggregator_output'] = Path.cwd().parent / configs['dev_config']['global']['relative_path']['aggregator_output']
configs['dev_config']['global']['relative_path']['processor_output'] = Path.cwd().parent / configs['dev_config']['global']['relative_path']['processor_output']

# Set reference dataset
datasetID = 'MiD17'

# Modify the localPathConfig file to point to the .csv file in the sampling folder in the tutorials directory where the dataset for the tutorials lies.
configs['user_config']['global']['absolute_path'][datasetID] = Path.cwd() /'data_sampling'

# Similarly we modify the datasetID in the global config file
configs['dev_config']['global']['files'][datasetID]['trips_data_raw'] = datasetID + '.csv'

# We also modify the parseConfig by removing some of the columns that are normally parsed from the MiD, which are not available in our semplified test dataframe
del configs['dev_config']['dataparsers']['data_variables']['household_id']
del configs['dev_config']['dataparsers']['data_variables']['person_id']


## DataParser config file

The DataParser config file defines which variables are to be parsed (i.e. the ones needed to create trip diaries and calculate fleet flexibility) and sets some filtering options, such as the conditions for trips to be included of excluded from the parsing.

<div class="alert alert-block alert-danger"><b>Warning:</b> The list is very long.</div>

In [19]:
configs['dev_config']

{'global': {'relative_path': {'parse_output': WindowsPath('c:/Users/mior_fa/Documents/7_Work/vencopy/vencopy_internal/vencopy/vencopy/output/dataparser'),
   'diary_output': WindowsPath('c:/Users/mior_fa/Documents/7_Work/vencopy/vencopy_internal/vencopy/vencopy/output/diarybuilder'),
   'grid_output': WindowsPath('c:/Users/mior_fa/Documents/7_Work/vencopy/vencopy_internal/vencopy/vencopy/output/gridmodeller'),
   'flex_output': WindowsPath('c:/Users/mior_fa/Documents/7_Work/vencopy/vencopy_internal/vencopy/vencopy/output/flexestimator'),
   'aggregator_output': WindowsPath('c:/Users/mior_fa/Documents/7_Work/vencopy/vencopy_internal/vencopy/vencopy/output/profileaggregator'),
   'processor_output': WindowsPath('c:/Users/mior_fa/Documents/7_Work/vencopy/vencopy_internal/vencopy/vencopy/output/postprocessor'),
   'config': './config/'},
  'files': {'MiD17': {'enrypted_zip_file_B1': 'B1_Standard-DatensatzpaketEncrypted.zip',
    'enrypted_zip_file_B2': 'B2_Regional-DatensatzpaketEncrypted.

## _DataParser_ class

Let's first run the class and see the outputs we get.

In [20]:
data = parse_data(configs=configs)
data.process()

Generic file parsing properties set up.
Starting to retrieve local data file from c:\Users\mior_fa\Documents\7_Work\vencopy\vencopy_internal\vencopy\vencopy\tutorials\data_sampling\MiD17.csv.
Finished loading 2124 rows of raw data of type .csv.
Running in debug mode.
Finished harmonization of variables.
Finished harmonization of ID variables.
Starting filtering, applying 8 filters.
All filters combined yielded that a total of 241 trips are taken into account.
This corresponds to 40.166666666666664 percent of the original data.
Completed park timestamp adjustments.
Finished activity composition with 240 trips and 240 parking activites.
Dataset written to c:\Users\mior_fa\Documents\7_Work\vencopy\vencopy_internal\vencopy\vencopy\output\dataparser\vencopy_output_dataparser_None_MiD17.csv.
Parsing MiD dataset completed.


index  unique_id  park_id  trip_id  is_driver  household_person_id  \
0        0       1410      0.0      NaN       True                 1410   
1        1       1410      NaN      4.0       True                 1410   
2        2       1410      4.0      NaN       True                 1410   
3        3        636      0.0      NaN       True                  636   
4        4        636      NaN      1.0       True                  636   
..     ...        ...      ...      ...        ...                  ...   
555    556       1840      0.0      NaN       True                 1840   
556    557       1840      NaN      1.0       True                 1840   
557    558       1840      1.0      NaN       True                 1840   
558    559       1840      NaN      2.0       True                 1840   
559    560       1840      2.0      NaN       True                 1840   

     trip_weight  trip_scale_factor  trip_purpose  trip_distance  ...  \
0       0.146627          39.299224             2            NaN  ...   
1       0.146627          39.299224             2           9.50  ...   
2       0.146627          39.299224             2            NaN  ...   
3       0.190006          50.925648             1            NaN  ...   
4       0.190006          50.925648             1          13.30  ...   
..           ...                ...           ...            ...  ...   
555     2.340718         627.363646             7            NaN  ...   
556     2.340718         627.363646             7           4.75  ...   
557     2.340718         627.363646             7            NaN  ...   
558     2.340718         627.363646             8           4.75  ...   
559     2.340718         627.363646             8            NaN  ...   

        timestamp_start       timestamp_end  is_first_activity  \
0   2017-03-13 00:00:00 2017-03-13 14:00:00               True   
1   2017-03-13 14:00:00 2017-03-13 14:50:00              False   
2   2017-03-13 14:50:00 2017-03-14 00:00:00              False   
3   2016-07-24 00:00:00 2016-07-24 06:10:00               True   
4   2016-07-24 06:10:00 2016-07-24 06:30:00              False   
..                  ...                 ...                ...   
555 2017-03-27 00:00:00 2017-03-27 09:45:00               True   
556 2017-03-27 09:45:00 2017-03-27 10:05:00              False   
557 2017-03-27 10:05:00 2017-03-27 11:30:00              False   
558 2017-03-27 11:30:00 2017-03-27 11:40:00              False   
559 2017-03-27 11:40:00 2017-03-28 00:00:00              False   

     is_last_activity activity_id next_activity_id previous_activity_id  \
0               False         0.0              4.0                  NaN   
1               False         4.0              4.0                  0.0   
2                True         4.0              NaN                  4.0   
3               False         0.0              1.0                  NaN   
4               False         1.0              1.0                  0.0   
..                ...         ...              ...                  ...   
555             False         0.0              1.0                  NaN   
556             False         1.0              1.0                  0.0   
557             False         1.0              2.0                  1.0   
558             False         2.0              2.0                  1.0   
559              True         2.0              NaN                  2.0   

    is_first_trip  is_first_park_activity      time_delta  
0           False                    True 0 days 14:00:00  
1            True                   False 0 days 00:50:00  
2           False                   False 0 days 09:10:00  
3           False                    True 0 days 06:10:00  
4            True                   False 0 days 00:20:00  
..            ...                     ...             ...  
555         False                    True 0 days 09:45:00  
556          True                   False 0 days 00:2

We can see from the print statements in the class that after reading in the initial dataset, which contained 2124 rows, and applying 8 filters, we end up with a database containing 1980 suitable entries, which corresponds to about 93% of the initial sample.
These trip respect the condition that they all need to be shorter than 1000km, which is set in the parseConfig under the 'filterDict' key.

Now we can, for example, change in the filters the maximum allowed trip distance from 1000km to 50km and see how this affects the resulting available trips (the extreme case of 50km is only used for the tutorial purpose).

In [21]:
configs['dev_config']['dataparsers']['filters'][datasetID]['smaller_than']['trip_distance'] = [50]

In [22]:
data = parse_data(configs=configs)
data.process()

Generic file parsing properties set up.
Starting to retrieve local data file from c:\Users\mior_fa\Documents\7_Work\vencopy\vencopy_internal\vencopy\vencopy\tutorials\data_sampling\MiD17.csv.
Finished loading 2124 rows of raw data of type .csv.
Running in debug mode.
Finished harmonization of variables.
Finished harmonization of ID variables.
Starting filtering, applying 8 filters.
All filters combined yielded that a total of 233 trips are taken into account.
This corresponds to 38.83333333333333 percent of the original data.
Completed park timestamp adjustments.
Finished activity composition with 232 trips and 232 parking activites.
Dataset written to c:\Users\mior_fa\Documents\7_Work\vencopy\vencopy_internal\vencopy\vencopy\output\dataparser\vencopy_output_dataparser_None_MiD17.csv.
Parsing MiD dataset completed.


index  unique_id  park_id  trip_id  is_driver  household_person_id  \
0        0       1410      0.0      NaN       True                 1410   
1        1       1410      NaN      4.0       True                 1410   
2        2       1410      4.0      NaN       True                 1410   
3        3        636      0.0      NaN       True                  636   
4        4        636      NaN      1.0       True                  636   
..     ...        ...      ...      ...        ...                  ...   
536    537       1840      0.0      NaN       True                 1840   
537    538       1840      NaN      1.0       True                 1840   
538    539       1840      1.0      NaN       True                 1840   
539    540       1840      NaN      2.0       True                 1840   
540    541       1840      2.0      NaN       True                 1840   

     trip_weight  trip_scale_factor  trip_purpose  trip_distance  ...  \
0       0.146627          39.299224             2            NaN  ...   
1       0.146627          39.299224             2           9.50  ...   
2       0.146627          39.299224             2            NaN  ...   
3       0.190006          50.925648             1            NaN  ...   
4       0.190006          50.925648             1          13.30  ...   
..           ...                ...           ...            ...  ...   
536     2.340718         627.363646             7            NaN  ...   
537     2.340718         627.363646             7           4.75  ...   
538     2.340718         627.363646             7            NaN  ...   
539     2.340718         627.363646             8           4.75  ...   
540     2.340718         627.363646             8            NaN  ...   

        timestamp_start       timestamp_end  is_first_activity  \
0   2017-03-13 00:00:00 2017-03-13 14:00:00               True   
1   2017-03-13 14:00:00 2017-03-13 14:50:00              False   
2   2017-03-13 14:50:00 2017-03-14 00:00:00              False   
3   2016-07-24 00:00:00 2016-07-24 06:10:00               True   
4   2016-07-24 06:10:00 2016-07-24 06:30:00              False   
..                  ...                 ...                ...   
536 2017-03-27 00:00:00 2017-03-27 09:45:00               True   
537 2017-03-27 09:45:00 2017-03-27 10:05:00              False   
538 2017-03-27 10:05:00 2017-03-27 11:30:00              False   
539 2017-03-27 11:30:00 2017-03-27 11:40:00              False   
540 2017-03-27 11:40:00 2017-03-28 00:00:00              False   

     is_last_activity activity_id next_activity_id previous_activity_id  \
0               False         0.0              4.0                  NaN   
1               False         4.0              4.0                  0.0   
2                True         4.0              NaN                  4.0   
3               False         0.0              1.0                  NaN   
4               False         1.0              1.0                  0.0   
..                ...         ...              ...                  ...   
536             False         0.0              1.0                  NaN   
537             False         1.0              1.0                  0.0   
538             False         1.0              2.0                  1.0   
539             False         2.0              2.0                  1.0   
540              True         2.0              NaN                  2.0   

    is_first_trip  is_first_park_activity      time_delta  
0           False                    True 0 days 14:00:00  
1            True                   False 0 days 00:50:00  
2           False                   False 0 days 09:10:00  
3           False                    True 0 days 06:10:00  
4            True                   False 0 days 00:20:00  
..            ...                     ...             ...  
536         False                    True 0 days 09:45:00  
537          True                   False 0 days 00:2

We can see how with a maximum trip distance of 1000km, all filters combined yielded a total of 1980 trips, which corresponds to about 93% of the original dataset. By changing this values to 50km, additional 82 trips have been excluded, resulting in 1898 trips (89% ofthe initial dataset).

## Next Steps

In the next tutorial, you will learn more in detail the internal workings of the FlexEstimator class and how to customise some settings.